pkg/api/customization/globalnamespaceaccess/access_common.go

```diff
@@ -299,116 +415,54 @@ func (ma *MemberAccess) getUserFromUserPrincipalID(userPrincipalID string) (*v3.
+	rolesToRemoveMap := make(map[string]bool)
+	if !removeAllRoles {
+		for _, role := range rolesToRemove {
+			rolesToRemoveMap[role] = true
 		}
-		for _, rt := range rts {
-			if rt.ClusterCreatorDefault && !clusterRolesToAddMap[rt.Name] {
-				clusterRolesToAddMap[rt.Name] = true
-			} else if rt.ProjectCreatorDefault && !projectRolesToAddMap[rt.Name] {
-				projectRolesToAddMap[rt.Name] = true
-			}
-		}
-		if len(projectRolesToAddMap) == 0 && len(clusterRolesToAddMap) == 0 {
-			return nil, httperror.NewAPIError(httperror.PermissionDenied, "Admin has passed no roles to multiclusterapp, and no cluster/project owner roles found")
-		}
-		// add projectCreatorDefault roles in all target projects
-		projectRolesToAdd := make([]string, len(projectRolesToAddMap))
-		i := 0
-		for role := range projectRolesToAddMap {
-			projectRolesToAdd[i] = role
-			i++
-		}
-		// add clusterCreatorDefault roles in all target projects's clusters
-		clusterRolesToAdd := make([]string, len(clusterRolesToAddMap))
-		i = 0
-		for role := range clusterRolesToAddMap {
-			clusterRolesToAdd[i] = role
-			i++
+	}
+
+	for _, target := range targetProjects {
+		split := strings.SplitN(target, ":", 2)
+		if len(split) != 2 {
+			errMsg := fmt.Sprintf("Invalid project ID: %v", target)
+			return httperror.NewAPIError(httperror.InvalidBodyContent, errMsg)
 		}
-		targetToRoles := make(map[string][]string)
-		for _, target := range targets {
-			targetToRoles[target] = projectRolesToAdd
-			split := strings.SplitN(target, ":", 2)
-			if len(split) != 2 {
-				return nil, fmt.Errorf("invalid project name: %v", target)
-			}
-			clusterName := split[0]
-			if _, ok := targetToRoles[clusterName]; !ok {
-				targetToRoles[clusterName] = clusterRolesToAdd
-			}
+		clusterName, projectName := split[0], split[1]
+		clustersCovered := make(map[string]bool)
+		prtbs, err := ma.PrtbLister.List(projectName, labels.NewSelector())
+		if err != nil {
+			return err
 		}
-	} else {
-		for _, target := range targets {
-			projectRoleFound := false
-			isClusterCreator := false
-			split := strings.SplitN(target, ":", 2)
-			if len(split) != 2 {
-				return targetToRoles, fmt.Errorf("invalid project name: %v", target)
-			}
-			clusterName := split[0]
-			projectName := split[1]
-			// get roles from this project for this creator
-			prtbs, err := ma.PrtbLister.List(projectName, labels.NewSelector())
-			if err != nil {
-				return targetToRoles, err
-			}
-			for _, prtb := range prtbs {
-				if prtb.UserName == callerID {
-					projectRoleFound = true
-					if roles, ok := targetToRoles[target]; ok {
-						if !slice.ContainsString(roles, prtb.RoleTemplateName) {
-							targetToRoles[target] = append(roles, prtb.RoleTemplateName)
-						}
-					} else {
-						targetToRoles[target] = []string{prtb.RoleTemplateName}
+		for _, prtb := range prtbs {
+			if prtb.UserPrincipalName == systemUserPrincipalID {
+				if (!removeAllRoles && rolesToRemoveMap[prtb.RoleTemplateName]) || removeAllRoles {
```

Comment:
You can just use:  if (removeAllRoles || rolesToRemoveMap[prtb.RoleTemplateName])?